### PR TITLE
[POAE7-2912] Comparison Operations Vectorization

### PR DESCRIPTION
**Pull request by bigPYJ1151**
### What changes were proposed in this pull request?

Comparison Operations Vectorization

### Why are the changes needed?

Performance optimization. 

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

UTs

### Which label does this PR belong to?

INFRA

Pls take a review @Ma-Jian1 @winningsix

### CODE REVIEW

File: cpp/src/cider/exec/nextgen/operators/RowToColumnNode.cpp [hunk 1]
```diff
@@ -329,21 +331,38 @@ class ColumnWriter {
...
     if (expr_->get_type_info().get_type() == kBOOLEAN) {
-      auto raw_data_buffer = context_.getJITFunction()->createLocalJITValue(
-          [this]() { return allocateBitwiseBuffer(1); });
+      auto raw_data_buffer = context_.getJITFunction()->createLocalJITValue([this]() {
+        if (!bitwise_bool_) {
+          // Memory padding for AVX2
+          JITValuePointer prev_len(arrow_array_len_);
+          auto padded_len = (prev_len + 31) / 32 * 32;
+          arrow_array_len_.replace(padded_len);
+          auto mem = allocateBitwiseBuffer(1);
```
Comment (Ma-Jian1) [inline cpp/src/cider/exec/nextgen/operators/RowToColumnNode.cpp:340]: `allocateBitwiseBuffer` add length argument.

File: cpp/src/cider/exec/nextgen/operators/VectorizedProjectNode.h [hunk 2]
```diff
@@ -39,10 +39,12 @@ class VectorizedProjectNode : public OpNode {
 
 class VectorizedProjectTranslator : public Translator {
  public:
+  enum class ExprsGroupType { BoolIOTree, BoolOutputTree, NormalTree, InvalidTree };
```
Comment (Ma-Jian1) [inline cpp/src/cider/exec/nextgen/operators/VectorizedProjectNode.h:42]: `InvalidTree` as the first unknown state?

File: cpp/src/cider/type/plan/Expr.h [hunk 3]
```diff
@@ -200,6 +200,8 @@ class Expr : public std::enable_shared_from_this<Expr> {
 
   bool isAutoVectorizable() const { return auto_vectorizable_; }
 
+  bool isColumnVar() const { return is_column_var_; }
```
Comment (Ma-Jian1) [inline cpp/src/cider/type/plan/Expr.h:203]: we always check type by dynamic_cast, why add this flag?

File: cpp/src/cider/util/CiderBitUtils.h [hunk 4]
```diff
@@ -240,6 +239,20 @@ inline void ENABLE_AVX256 bitwiseAnd(uint8_t* __restrict output,
...
 }
+
+inline void ENABLE_AVX256 byteToBit(uint8_t* __restrict byte,
+                                    uint8_t* __restrict bit,
+                                    size_t bit_num) {
+  __m256i* byte_256 = (__m256i*)byte;
+  int* bit_32 = (int*)bit;
+  bit_num += 31;
+  bit_num >>= 5;
+  for (size_t i = 0; i < bit_num; ++i) {
```
Comment (Ma-Jian1) [inline cpp/src/cider/util/CiderBitUtils.h:250]: can't normal loop auto-vectorization by the compiler?

Comment (bigPYJ1151) [reply cpp/src/cider/util/CiderBitUtils.h:250]: Yes, the compiler can't handle the vectorization of bit operations.

File: cpp/src/cider/type/plan/Expr.h (hunk 3 above)
Comment (bigPYJ1151) [reply cpp/src/cider/type/plan/Expr.h:203]: Reduce RTTI cost.

File: cpp/src/cider/exec/nextgen/operators/RowToColumnNode.cpp (hunk 1 above)
Comment (bigPYJ1151) [reply cpp/src/cider/exec/nextgen/operators/RowToColumnNode.cpp:340]: Some interfaces of the writer and reader need to be decoupled.

**Review by Ma-Jian1 — APPROVED** [review]

**Review by harborn — APPROVED** [review]

**Pull request merged**